### DISCUSSION TITLE
Module federation optimization - all remote entries contains vendor code

### DISCUSSION THREAD

**Discussion by sylhero** [webpack/webpack#16659] [Q&A]
Hello,

I have a simple setup like shell -> appRoute -> libs. When I generate the remote entries I have appRouteRemoteEntry.js and libsRemoteEntry.js but when I use the bundle analyzer I see both remote entries contain vendor code like react, react-dom. I also used the `shared` feature module federation. My question is could we specify the run time chunk and load the `shared` vendor code beforehand so each remote entry can only contain the source code? 

```
 optimization: isProduction
    ? {
        chunkIds: 'size',
        // method of generating ids for chunks
        moduleIds: 'size',
        // method of generating ids for modules
        mangleExports: 'size',
        // rename export names to shorter names
        runtimeChunk: 'single',

        minimize: true,
        // minimize the output files
        minimizer: [
          new TerserPlugin({
            terserOptions: {
              format: {
                comments: isProduction ? false : 'all',
              },
            },
            extractComments: false, // do not extract comments to the *.LICENSE.txt file
          }),
          new CssMinimizerPlugin(),
        ],
        // minimizers to use for the output files
      }
    : {},
```

**Comment by leungtinyau27m5** [comment]
Great question! I stuck here too, the JS size (without tree shaking) is actually big not "micro" enough.

**Reply by alexander-akait** [reply]
> My question is could we specify the run time chunk and load the shared vendor code beforehand so each remote entry can only contain the source code?

It should work as you describe, need to look at code to say what is wrong, can you create minimum reproducible test repo to provide example such behaviour? Thank you

**Comment by joebnb** [comment]
did this was solved? use module federation and check output,i was also confusing about bundle size,it's seems not exclude dependencies from MF configuration


eg:  
```js
new ModuleFederationPlugin({
      name: 'hostApp',
      remotes: {
        remoteApp: 'remoteApp@static/remoteApp.js',
      },
      shared: {
        '@zenlayer/zen-design': { singleton: true },
        '@zenlayer/zen-design-pro': { singleton: true },
        antd: { singleton: true },
      },
    }),
```


```js
//host
new ModuleFederationPlugin({
    name: 'remoteApp',
    library: { type: 'umd', name: 'remoteApp' },
    filename: 'static/remoteApp.js',
    exposes: {
      './entry': './src/remoteEntry',
    },
    shared: {
      react: { singleton: true, eager: true },
      'react-dom': { singleton: true, eager: true },
      antd: { singleton: true },
    },
  }),

```

in this configuration webpack will bundle dependecies each project and remoteEntry:
react and react-dom in: host app,host entry,remote app
antd also in : host app,host entry,remote app

why caused this，what is the essential condition of module federation exclude bundld and how i check it
